DataManagementSystem/Utilities/CSHelpers.py

```diff
@@ -0,0 +1,235 @@
+    """
+    if self.siteSEMapping:
+      return S_OK( self.siteSEMapping )
+
+    siteSEMapping = {}
+    gridTypes = gConfig.getSections( 'Resources/Sites/' )
+    if not gridTypes['OK']:
+      gLogger.warn( 'Problem retrieving sections in /Resources/Sites' )
+      return gridTypes
+
+    gridTypes = gridTypes['Value']
+
+    gLogger.debug( 'Grid Types are: %s' % ( ', '.join( gridTypes ) ) )
+    # Get a list of sites and their local SEs
+    siteSet = set()
+    storageElementSet = set()
+    siteSEMapping[self.local] = {}
+    for grid in gridTypes:
+      result = gConfig.getSections( '/Resources/Sites/%s' % grid )
+      if not result['OK']:
+        gLogger.warn( 'Problem retrieving /Resources/Sites/%s section' % grid )
+        return result
+      sites = result['Value']
+      siteSet.update( sites )
+      for site in sites:
+        candidateSEs = gConfig.getValue( '/Resources/Sites/%s/%s/SE' % ( grid, site ), [] )
+        if candidateSEs:
+          siteSEMapping[self.local].setdefault( site, set() ).update( candidateSEs )
+          storageElementSet.update( candidateSEs )
+        else:
+          gLogger.debug( 'No SEs defined for site %s' % site )
+
+    # Add Sites from the SiteSEMappingByProtocol in the CS
+    siteSEMapping[self.protocol] = {}
+    cfgLocalSEPath = cfgPath( 'SiteSEMappingByProtocol' )
+    result = self.__opsHelper.getOptionsDict( cfgLocalSEPath )
+    if result['OK']:
+      sites = result['Value']
+      for site in sites:
+        candidates = self.__opsHelper.getValue( cfgPath( cfgLocalSEPath, site ), [] )
+        ses = set( candidates )
+        # If a candidate is a site, then all local SEs are eligible
+        for candidate in ses & siteSet:
+          ses.remove( candidate )
+          ses.update( siteSEMapping[self.local][candidate] )
+        siteSEMapping[self.protocol].setdefault( site, set() ).update( ses )
+
+    # Add Sites from the SiteSEMappingByDownload in the CS, else SiteLocalSEMapping (old convention)
+    siteSEMapping[self.download] = {}
+    cfgLocalSEPath = cfgPath( 'SiteSEMappingByDownload' )
+    result = self.__opsHelper.getOptionsDict( cfgLocalSEPath )
+    if not result['OK']:
+      cfgLocalSEPath = cfgPath( 'SiteLocalSEMapping' )
+      result = self.__opsHelper.getOptionsDict( cfgLocalSEPath )
+    if result['OK']:
+      sites = result['Value']
+      for site in sites:
+        candidates = self.__opsHelper.getValue( cfgPath( cfgLocalSEPath, site ), [] )
+        ses = set( candidates )
+        for candidate in ses & siteSet:
+          ses.remove( candidate )
+          ses.update( siteSEMapping[self.local][candidate] )
+        siteSEMapping[ self.download].setdefault( site, set() ).update( ses )
+
+    self.siteSEMapping = siteSEMapping
+    self.storageElementSet = storageElementSet
+    self.siteSet = siteSet
+    return S_OK( siteSEMapping )
+
+
+  def isSEFailover( self, storageElement ):
```

Comment:
This new CS param should be mentioned in the wiki 'how to update', and then we can use it in the Request.Optimzie code :-)